services/QuillLMS/app/services/analytics/segment_analytics.rb

```diff
@@ -71,12 +71,36 @@ def track_activity_pack_assignment(teacher_id, unit_id)
     })
   end
 
-  def track_activity_completion(user, student_id, activity)
+  def track_activity_completion(user, student_id, activity, activity_session)
     track({
       user_id: user&.id,
       event: SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION,
       properties: activity.segment_activity.content_params.merge({student_id: student_id})
     })
+    track_activity_pack_completion(user, student_id, activity_session) if activity_pack_completed?(student_id, activity_session)
+  end
+
+  def activity_pack_completed?(student_id, activity_session)
+    classroom_unit = ClassroomUnit.find_by(id: activity_session.classroom_unit_id)
+    activity_ids = classroom_unit&.unit_activities&.pluck(:activity_id)
+    incomplete_session = activity_ids.find do |activity_id|
+      completed_activity_session = ActivitySession.find_by(user_id: student_id, activity_id: activity_id, state: "finished", classroom_unit: classroom_unit)
+      completed_activity_session.nil?
+    end
+    incomplete_session.nil?
+  end
+
+  def track_activity_pack_completion(user, student_id, activity_session)
+    unit = ClassroomUnit.find(activity_session.classroom_unit_id)&.unit
```

Comment:
Activity sessions have a `has_one` relationship with units, so you can just do `unit = activity_session.unit` here.